pkg/gadgets/common/filesystem.h

```diff
@@ -0,0 +1,209 @@
+                bpf_probe_read(&mnt_p, sizeof(struct mount *), &mnt_p->mnt_parent);
+                bpf_probe_read(&mnt_parent_p, sizeof(struct mount *), &mnt_p->mnt_parent);
+                vfsmnt = &mnt_p->mnt;
+                continue;
+            }
+            // Global root - path fully parsed
+            break;
+        }
+        // Add this dentry name to path
+        d_name = get_d_name_from_dentry(dentry);
+        len = (d_name.len + 1) & (MAX_STRING_SIZE - 1);
+        off = buf_off - len;
+
+        // Is string buffer big enough for dentry name?
+        sz = 0;
+        if (off <= buf_off) { // verify no wrap occurred
+            len = len & ((MAX_PERCPU_BUFSIZE >> 1) - 1);
+            sz = bpf_probe_read_str(
+                &(string_p->buf[off & ((MAX_PERCPU_BUFSIZE >> 1) - 1)]), len, (void *) d_name.name);
+        } else
+            break;
+        if (sz > 1) {
+            buf_off -= 1; // remove null byte termination with slash sign
+            bpf_probe_read(&(string_p->buf[buf_off & (MAX_PERCPU_BUFSIZE - 1)]), 1, &slash);
+            buf_off -= sz - 1;
+        } else {
+            // If sz is 0 or 1 we have an error (path can't be null nor an empty string)
+            break;
+        }
+        dentry = d_parent;
+    }
+
+    if (buf_off == (MAX_PERCPU_BUFSIZE >> 1)) {
+        // memfd files have no path in the filesystem -> extract their name
+        buf_off = 0;
+        d_name = get_d_name_from_dentry(dentry);
+        bpf_probe_read_str(&(string_p->buf[0]), MAX_STRING_SIZE, (void *) d_name.name);
+    } else {
+        // Add leading slash
+        buf_off -= 1;
+        bpf_probe_read(&(string_p->buf[buf_off & (MAX_PERCPU_BUFSIZE - 1)]), 1, &slash);
+        // Null terminate the path string
+        bpf_probe_read(&(string_p->buf[(MAX_PERCPU_BUFSIZE >> 1) - 1]), 1, &zero);
+    }
+
+    return &string_p->buf[buf_off];
+}
+
+// Function to extract file structure from a user space file descriptor
+static __always_inline struct file * get_struct_file_for_fd(int fd_num)
+{
+    if (fd_num < 0) {
+        return NULL;
+    }
+    
+    struct task_struct *task = (struct task_struct *) bpf_get_current_task();
+    if (task == NULL) {
+        return NULL;
+    }
+
+	// extract the file vector from the task_struct
+	struct file **fd = BPF_CORE_READ(task, files, fdt, fd);
+	
+	// extract the file pointer from the file vector
+	struct file *f = NULL;
+    uint max_fds = BPF_CORE_READ(task, files, fdt, max_fds);
+    if (fd_num < max_fds) {
+	    bpf_core_read((void *) &f, sizeof(f), &fd[fd_num]);
+    }
+
+	return f;
+}
+
+static __always_inline int read_full_path_of_open_file_fd(int fd_num, char *buf, u64 buf_len)
+{
+	struct file *file = get_struct_file_for_fd(fd_num);
+	if (file == NULL) {
+		return 1;
+	}
+
+	struct path f_path = BPF_CORE_READ(file, f_path);
+
+	// Extract the full path string
+	char* c_path = get_path_str(__builtin_preserve_access_index(&f_path));
```

Comment:
We use vmlinux.h that marks all structs with [preserve_access_index](https://github.com/alban/inspektor-gadget/blob/d135fa553e5a6957b9aee2274762db55b36db517/pkg/amd64/vmlinux/vmlinux.h#L5), so we don't need to make this manually. We can also use `get_path_str(&file->f_path)` directly to avoid creating the temp variable.

Reply:
gotcha